sklearn/cross_validation.py

```diff
@@ -332,6 +332,112 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(y, n_folds=3):
+    """Creates folds where a same label is not in two different folds
+    
+    Parameters
+    ----------
+    y: iterable of shape (n_samples, )
+        contains an id for each sample
+        The folds are built so that the same id doesn't appear in two different folds
+    
+    n_folds: int, default is 3
+        number of folds to split the data into
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        array of integers between 0 and (n_folds - 1)
+        folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    """
+    labels = np.array(y)
+    unique_labels = np.unique(labels)
+    
+    # number of occurrence of each label (its "weight")
+    weight_per_label = sorted([(sum(labels == label), label) for label in unique_labels])
+    # Total weight of each fold
+    weight_per_fold = np.zeros(n_folds)
+    # For each sample, a digit between 0 and (n_folds - 1) to tell which fold it belongs to
+    folds = np.zeros(len(labels))
+    
+    # While there are weights, distribute them
+    # Specifically, add the biggest weight to the lightest fold
+    while weight_per_label:
+        ind_min = np.argmin(weight_per_fold)
+        w, label = weight_per_label.pop()
+        weight_per_fold[ind_min] += w
+        folds[labels == label] = ind_min
```

Comment:
maybe instead of filling the folds as you go, just keep an array ``label_to_fold`` of length ``unique_labels`` that maps for each label which fold it will go to, and later do ``folds = label_to_fold[y]``